File: lib/code_puppy_control/plugins/agent_trace/schema.ex
defmodule CodePuppyControl.Plugins.AgentTrace.Schema do
  @moduledoc """
  Schema definitions for agent trace events (V2).

  Defines the core data structures: NodeKind, TransferKind, EventType,
  TraceEvent, and supporting structs. These are the atoms of the
  execution graph that the reducer assembles into TraceState.
  """

  @type node_kind ::
          :user | :session | :agent_run | :model_call | :tool_call | :memory_snapshot | :artifact
  @type transfer_kind ::
          :user_prompt
          | :system_instructions
          | :history_context
          | :retrieved_context
          | :model_input
          | :model_output
          | :tool_args
          | :tool_result
          | :delegate_prompt
          | :delegate_response
          | :memory_append
          | :artifact_write
          | :artifact_read
  @type token_class ::
          :input_tokens
          | :output_tokens
          | :reasoning_tokens
          | :cached_tokens
          | :estimated_tokens
          | :billable_tokens
  @type accounting_state :: :estimated_live | :provider_reported_exact | :reconciled | :unknown
  @type event_type ::
          :span_started
          | :span_updated
          | :span_ended
          | :transfer_started
          | :transfer_chunk
          | :transfer_completed
          | :usage_reported
          | :usage_reconciled
          | :artifact_created
          | :artifact_read

  @type trace_event :: %{
          event_id: String.t(),
          trace_id: String.t(),
          event_type: event_type(),
          timestamp: String.t(),
          span_id: String.t() | nil,
          parent_span_id: String.t() | nil,
          run_id: String.t() | nil,
          session_id: String.t() | nil,
          node: map() | nil,
          transfer: map() | nil,
          metrics: map() | nil,
          extra: map()
        }

  @doc "Generate a unique event ID."
  @spec make_event_id() :: String.t()
  def make_event_id do
    :crypto.strong_rand_bytes(16) |> Base.encode16(case: :lower)
  end

  @doc "Get current UTC timestamp in ISO format."
  @spec now_iso() :: String.t()
  def now_iso, do: DateTime.utc_now() |> DateTime.to_iso8601()

  @doc "Create a span ID."
  @spec make_span_id() :: String.t()
  def make_span_id do
    "span-#{:crypto.strong_rand_bytes(6) |> Base.encode16(case: :lower)}"
  end

  @doc "Create a node ID with a meaningful prefix."
  @spec make_node_id(atom(), String.t() | nil) :: String.t()
  def make_node_id(kind, name \\ nil) do
    suffix = :crypto.strong_rand_bytes(4) |> Base.encode16(case: :lower)

    if name do
      safe_name = name |> String.replace(~r/[^a-zA-Z0-9]/, "-") |> String.slice(0, 20)
      "#{kind}-#{safe_name}-#{suffix}"
    else
      "#{kind}-#{suffix}"
    end
  end

  @doc "Create a span.started event."
  @spec span_started(String.t(), atom(), keyword()) :: trace_event()
  def span_started(trace_id, kind, opts \\ []) do
    span_id = Keyword.get(opts, :span_id, make_span_id())
    name = Keyword.get(opts, :name)
    parent_span_id = Keyword.get(opts, :parent_span_id)
    session_id = Keyword.get(opts, :session_id)
    run_id = Keyword.get(opts, :run_id)
    parent_node_id = Keyword.get(opts, :parent_node_id)
    node_id = make_node_id(kind, name)

    %{
      event_id: make_event_id(),
      trace_id: trace_id,
      event_type: :span_started,
      timestamp: now_iso(),
      span_id: span_id,
      parent_span_id: parent_span_id,
      run_id: run_id,
      session_id: session_id,
      node: %{
        id: node_id,
        kind: kind,
        name: name,
        status: "running",
        parent_node_id: parent_node_id
      },
      transfer: nil,
      metrics: nil,
      extra: %{}
    }
  end

  @doc "Create a span.ended event."
  @spec span_ended(String.t(), String.t(), String.t(), atom(), keyword()) :: trace_event()
  def span_ended(trace_id, span_id, node_id, kind, opts \\ []) do
    name = Keyword.get(opts, :name)
    success = Keyword.get(opts, :success, true)
    error = Keyword.get(opts, :error)
    duration_ms = Keyword.get(opts, :duration_ms)
    session_id = Keyword.get(opts, :session_id)

    %{
      event_id: make_event_id(),
      trace_id: trace_id,
      event_type: :span_ended,
      timestamp: now_iso(),
      span_id: span_id,
      parent_span_id: nil,
      run_id: nil,
      session_id: session_id,
      node: %{
        id: node_id,
        kind: kind,
        name: name,
        status: if(success, do: "done", else: "failed")
      },
      transfer: nil,
      metrics: if(duration_ms, do: %{duration_ms: duration_ms}, else: nil),
      extra: if(error, do: %{error: to_string(error)}, else: %{})
    }
  end

  @doc "Create a usage.reconciled event."
  @spec usage_reconciled(String.t(), String.t(), String.t(), keyword()) :: trace_event()
  def usage_reconciled(trace_id, span_id, node_id, opts \\ []) do
    session_id = Keyword.get(opts, :session_id)
    est_input = Keyword.get(opts, :estimated_input)
    est_output = Keyword.get(opts, :estimated_output)
    exact_input = Keyword.get(opts, :exact_input)
    exact_output = Keyword.get(opts, :exact_output)
    exact_reasoning = Keyword.get(opts, :exact_reasoning)
    exact_cached = Keyword.get(opts, :exact_cached)
    cost_usd = Keyword.get(opts, :cost_usd)

    reconciliation = %{
      node_id: node_id,
      estimated: %{input_tokens: est_input, output_tokens: est_output},
      exact: %{
        input_tokens: exact_input,
        output_tokens: exact_output,
        reasoning_tokens: exact_reasoning,
        cached_tokens: exact_cached
      },
      drift: %{
        input_tokens: if(exact_input && est_input, do: exact_input - est_input),
        output_tokens: if(exact_output && est_output, do: exact_output - est_output)
      }
    }

    %{
      event_id: make_event_id(),
      trace_id: trace_id,
      event_type: :usage_reconciled,
      timestamp: now_iso(),
      span_id: span_id,
      parent_span_id: nil,
      run_id: nil,
      session_id: session_id,
      node: nil,
      transfer: %{
        kind: :model_output,
        source_node_id: node_id,
        token_count: exact_output,
        token_class: :output_tokens,
        accounting: :reconciled
      },
      metrics: if(cost_usd, do: %{cost_usd: cost_usd}, else: nil),
      extra: %{reconciliation: reconciliation}
    }
  end

  @doc "Serialize a trace event to JSON."
  @spec to_json(trace_event()) :: String.t()
  def to_json(event) do
    Jason.encode!(event)
  end

  @doc "Deserialize a trace event from JSON."
  @spec from_json(String.t()) :: trace_event()
  def from_json(json_str) do
    Jason.decode!(json_str, keys: :atoms)
  end
end

File: code_puppy/plugins/agent_trace/schema.py
"""Agent Trace V2 — Normalized Event Schema.

Based on agent_flow_v2_schema_example.json. This module defines:
- Event types for spans, transfers, and usage
- Node kinds: user, agent_run, model_call, tool_call, etc.
- Transfer kinds: model_input, model_output, tool_args, tool_result, etc.
- Token classes with accounting state (estimated vs exact)
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class NodeKind(str, Enum):
    """Types of nodes in the execution graph."""

    USER = "user"
    SESSION = "session"
    AGENT_RUN = "agent_run"
    MODEL_CALL = "model_call"
    TOOL_CALL = "tool_call"
    MEMORY_SNAPSHOT = "memory_snapshot"
    ARTIFACT = "artifact"


class TransferKind(str, Enum):
    """Types of token/data transfers between nodes."""

    USER_PROMPT = "user_prompt"
    SYSTEM_INSTRUCTIONS = "system_instructions"
    HISTORY_CONTEXT = "history_context"
    RETRIEVED_CONTEXT = "retrieved_context"
    MODEL_INPUT = "model_input"
    MODEL_OUTPUT = "model_output"
    TOOL_ARGS = "tool_args"
    TOOL_RESULT = "tool_result"
    DELEGATE_PROMPT = "delegate_prompt"
    DELEGATE_RESPONSE = "delegate_response"
    MEMORY_APPEND = "memory_append"
    ARTIFACT_WRITE = "artifact_write"
    ARTIFACT_READ = "artifact_read"


class TokenClass(str, Enum):
    """Categories of tokens for accounting."""

    INPUT_TOKENS = "input_tokens"
    OUTPUT_TOKENS = "output_tokens"
    REASONING_TOKENS = "reasoning_tokens"
    CACHED_TOKENS = "cached_tokens"
    ESTIMATED_TOKENS = "estimated_tokens"
    BILLABLE_TOKENS = "billable_tokens"


class AccountingState(str, Enum):
    """Confidence level of token counts.

    This is the key insight from V2: separate what we know for sure
    from what we're estimating live.
    """

    ESTIMATED_LIVE = "estimated_live"  # Calculated from stream chunks
    PROVIDER_REPORTED_EXACT = "provider_reported_exact"  # From API usage response
    RECONCILED = "reconciled"  # Estimate corrected by exact value
    UNKNOWN = "unknown"  # No information available


class EventType(str, Enum):
    """Types of trace events."""

    SPAN_STARTED = "span.started"
    SPAN_UPDATED = "span.updated"
    SPAN_ENDED = "span.ended"
    TRANSFER_STARTED = "transfer.started"
    TRANSFER_CHUNK = "transfer.chunk"
    TRANSFER_COMPLETED = "transfer.completed"
    USAGE_REPORTED = "usage.reported"
    USAGE_RECONCILED = "usage.reconciled"
    ARTIFACT_CREATED = "artifact.created"
    ARTIFACT_READ = "artifact.read"


@dataclass
class NodeInfo:
    """Information about a node in the execution graph."""

    id: str
    kind: NodeKind
    name: str | None = None
    status: str | None = None
    parent_node_id: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "kind": self.kind.value,
            "name": self.name,
            "status": self.status,
            "parent_node_id": self.parent_node_id,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> NodeInfo:
        return cls(
            id=data["id"],
            kind=NodeKind(data["kind"]),
            name=data.get("name"),
            status=data.get("status"),
            parent_node_id=data.get("parent_node_id"),
        )


@dataclass
class TransferInfo:
    """Information about a token/data transfer between nodes."""

    kind: TransferKind
    source_node_id: str | None = None
    target_node_id: str | None = None
    message_id: str | None = None
    token_count: int | None = None
    token_class: TokenClass | None = None
    accounting: AccountingState = AccountingState.UNKNOWN
    preview: str | None = None  # Truncated/redacted content preview

    def to_dict(self) -> dict[str, Any]:
        return {
            "kind": self.kind.value,
            "source_node_id": self.source_node_id,
            "target_node_id": self.target_node_id,
            "message_id": self.message_id,
            "token_count": self.token_count,
            "token_class": self.token_class.value if self.token_class else None,
            "accounting": self.accounting.value,
            "preview": self.preview,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> TransferInfo:
        return cls(
            kind=TransferKind(data["kind"]),
            source_node_id=data.get("source_node_id"),
            target_node_id=data.get("target_node_id"),
            message_id=data.get("message_id"),
            token_count=data.get("token_count"),
            token_class=TokenClass(data["token_class"])
            if data.get("token_class")
            else None,
            accounting=AccountingState(data.get("accounting", "unknown")),
            preview=data.get("preview"),
        )


@dataclass
class MetricsInfo:
    """Timing and cost metrics for a span or transfer."""

    duration_ms: float | None = None
    cost_usd: float | None = None
    queue_time_ms: float | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "duration_ms": self.duration_ms,
            "cost_usd": self.cost_usd,
            "queue_time_ms": self.queue_time_ms,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> MetricsInfo:
        return cls(
            duration_ms=data.get("duration_ms"),
            cost_usd=data.get("cost_usd"),
            queue_time_ms=data.get("queue_time_ms"),
        )


def _generate_id() -> str:
    """Generate a unique event ID."""
    return str(uuid.uuid4())


def _now_iso() -> str:
    """Get current UTC timestamp in ISO format."""
    return datetime.now(timezone.utc).isoformat()


@dataclass
class TraceEvent:
    """A normalized trace event.

    This is the core data structure for Agent Flow V2. Every observation
    about agent execution is represented as a TraceEvent.
    """

    event_id: str = field(default_factory=_generate_id)
    trace_id: str = ""  # Root session/run identifier
    event_type: EventType = EventType.SPAN_STARTED
    timestamp: str = field(default_factory=_now_iso)

    # Optional identifiers for correlation
    span_id: str | None = None
    parent_span_id: str | None = None
    run_id: str | None = None
    session_id: str | None = None

    # Structured data
    node: NodeInfo | None = None
    transfer: TransferInfo | None = None
    metrics: MetricsInfo | None = None

    # Extension point for additional data
    extra: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        """Serialize to dictionary for JSON output."""
        result = {
            "event_id": self.event_id,
            "trace_id": self.trace_id,
            "event_type": self.event_type.value,
            "timestamp": self.timestamp,
        }

        if self.span_id:
            result["span_id"] = self.span_id
        if self.parent_span_id:
            result["parent_span_id"] = self.parent_span_id
        if self.run_id:
            result["run_id"] = self.run_id
        if self.session_id:
            result["session_id"] = self.session_id
        if self.node:
            result["node"] = self.node.to_dict()
        if self.transfer:
            result["transfer"] = self.transfer.to_dict()
        if self.metrics:
            result["metrics"] = self.metrics.to_dict()
        if self.extra:
            result["extra"] = self.extra

        return result

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> TraceEvent:
        """Deserialize from dictionary."""
        return cls(
            event_id=data["event_id"],
            trace_id=data["trace_id"],
            event_type=EventType(data["event_type"]),
            timestamp=data["timestamp"],
            span_id=data.get("span_id"),
            parent_span_id=data.get("parent_span_id"),
            run_id=data.get("run_id"),
            session_id=data.get("session_id"),
            node=NodeInfo.from_dict(data["node"]) if data.get("node") else None,
            transfer=TransferInfo.from_dict(data["transfer"])
            if data.get("transfer")
            else None,
            metrics=MetricsInfo.from_dict(data["metrics"])
            if data.get("metrics")
            else None,
            extra=data.get("extra", {}),
        )

    def to_json(self) -> str:
        """Serialize to JSON string."""
        import json

        return json.dumps(self.to_dict())

    @classmethod
    def from_json(cls, json_str: str) -> TraceEvent:
        """Deserialize from JSON string."""
        import json

        return cls.from_dict(json.loads(json_str))
